src/Middleware/LazyLoadingMiddleware.php

```diff
@@ -0,0 +1,82 @@
+<?php
+/**
+ * @see       https://github.com/zendframework/zend-expressive for the canonical source repository
+ * @copyright Copyright (c) 2017 Zend Technologies USA Inc. (http://www.zend.com)
+ * @license   https://github.com/zendframework/zend-expressive/blob/master/LICENSE.md New BSD License
+ */
+
+namespace Zend\Expressive\Middleware;
+
+use Interop\Container\ContainerInterface;
+use Interop\Http\ServerMiddleware\DelegateInterface;
+use Interop\Http\ServerMiddleware\MiddlewareInterface as ServerMiddlewareInterface;
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+use Zend\Expressive\Exception\InvalidMiddlewareException;
+use Zend\Expressive\IsCallableInteropMiddlewareTrait;
+
+class LazyLoadingMiddleware implements ServerMiddlewareInterface
+{
+    use IsCallableInteropMiddlewareTrait;
+
+    /**
+     * @var ContainerInterface
+     */
+    private $container;
+
+    /**
+     * @var string
+     */
+    private $middlewareName;
+
+    /**
+     * @var ResponseInterface
+     */
+    private $responsePrototype;
+
+    public function __construct(
+        ContainerInterface $container,
+        ResponseInterface $responsePrototype,
+        $middlewareName
+    ) {
+        $this->container = $container;
+        $this->responsePrototype = $responsePrototype;
+        $this->middlewareName = $middlewareName;
+    }
+
+    /**
+     * @param ServerRequestInterface $request
+     * @param DelegateInterface $delegate
+     * @return ResponseInterface
+     * @throws InvalidMiddlewareException for invalid middleware types pulled
+     *     from the container.
+     */
+    public function process(ServerRequestInterface $request, DelegateInterface $delegate)
+    {
+        $middleware = $this->container->get($this->middlewareName);
+
+        // http-interop middleware
+        if ($middleware instanceof ServerMiddlewareInterface) {
+            return $middleware->process($request, $delegate);
+        }
+
+        // Unknown - invalid!
+        if (! is_callable($middleware)) {
+            throw new InvalidMiddlewareException(sprintf(
+                'Lazy-loaded middleware "%s" is neither invokable nor implements %s',
+                $this->middlewareName,
+                ServerMiddlewareInterface::class
+            ));
+        }
+
+        // Callable http-interop middleware
+        if ($this->isCallableInteropMiddleware($middleware)) {
```

Comment:
Are there enough users for this "edge case" to be included in a major version change? It would seem that this is targeting a rather small/nonexistent(?) percentage of users that a) have http-interop style middleware and b) use `__invoke()` instead of `process()`. Is this more to allow users to include older 3rd-party http-interop middleware?